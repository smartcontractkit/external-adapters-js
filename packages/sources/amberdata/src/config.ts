import { Requester } from '@chainlink/ea-bootstrap'
import { Config } from '@chainlink/types'

export const NAME = 'AMBERDATA'

export const DEFAULT_API_ENDPOINT = 'https://web3api.io'
export const DEFAULT_WS_API_ENDPOINT = 'wss://ws.web3api.io'

export const COIN_KEYS = ['btc', 'eth', 'bch', 'ltc', 'btsv', 'zec'] as const
export type CoinType = typeof COIN_KEYS[number]
export function isCoinType(key: string): key is CoinType {
  return COIN_KEYS.includes(key as CoinType)
}
export const CHAIN_KEYS = ['mainnet'] as const
export type ChainType = typeof CHAIN_KEYS[number]
export function isChainType(key: string): key is ChainType {
  return CHAIN_KEYS.includes(key as ChainType)
}

export const BLOCKCHAINS: { [ticker: string]: string } = {
  btc: 'bitcoin',
  eth: 'ethereum',
  bch: 'bitcoin-abc',
  ltc: 'litecoin',
  btsv: 'bitcoin-sv',
  zec: 'zcash',
}

const DEFAULT_ENDPOINT = 'price'

export const makeConfig = (prefix = ''): Config => {
  const config = Requester.getDefaultConfig(prefix, true)
  config.api.headers['x-api-key'] = config.apiKey
  config.api.baseURL = config.api.baseURL || DEFAULT_API_ENDPOINT
  config.defaultEndpoint = DEFAULT_ENDPOINT
  return config
}
