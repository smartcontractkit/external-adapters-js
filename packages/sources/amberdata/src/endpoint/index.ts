export * as balance from './balance'
export * as crypto from './crypto'
export * as token from './token'
export * as gasprice from './gasprice'
