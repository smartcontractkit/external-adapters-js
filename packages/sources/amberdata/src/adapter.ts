import { Requester, Validator, Builder } from '@chainlink/ea-bootstrap'
import {
  AdapterRequest,
  Config,
  ExecuteFactory,
  ExecuteWithConfig,
  MakeWSHandler,
  APIEndpoint,
} from '@chainlink/types'
import { DEFAULT_WS_API_ENDPOINT, makeConfig, NAME } from './config'
import * as endpoints from './endpoint'
import { crypto } from './endpoint'

// Export function to integrate with Chainlink node
export const execute: ExecuteWithConfig<Config> = async (request, context, config) => {
  return Builder.buildSelector(request, context, config, endpoints)
}

export const endpointSelector = (request: AdapterRequest): APIEndpoint =>
  Builder.selectEndpoint(request, makeConfig(), endpoints)

export const makeExecute: ExecuteFactory<Config> = (config) => {
  return async (request, context) => execute(request, context, config || makeConfig())
}

export const makeWSHandler = (defaultConfig?: Config): MakeWSHandler => {
  const subscriptions: any = {}
  const getPair = (input: AdapterRequest) => {
    const validator = new Validator(input, crypto.inputParameters, {}, false)
    if (validator.error) return
    const base = (validator.overrideSymbol(NAME) as string).toLowerCase()
    const quote = validator.validated.data.quote.toLowerCase()
    return `${base}_${quote}`
  }
  const getSubscription = (pair?: string) => {
    if (!pair) return
    return { id: 1, method: 'subscribe', params: ['market:tickers', { pair }] }
  }
  const getUnsubscription = (pair?: string) => {
    if (!pair) return
    return { id: 1, method: 'unsubscribe', params: [subscriptions[pair]] }
  }
  return () => {
    const config = defaultConfig || makeConfig()
    return {
      connection: {
        url: config.api.baseWsURL || DEFAULT_WS_API_ENDPOINT,
        protocol: { headers: { ...config.api.headers } },
      },
      subscribe: (input) => getSubscription(getPair(input)),
      unsubscribe: (input) => getUnsubscription(getPair(input)),
      subsFromMessage: (message) => {
        const pair = message?.params?.result?.pair
        subscriptions[pair] = message?.params?.subscription
        return getSubscription(message?.params?.result?.pair)
      },
      // https://github.com/web3data/web3data-js/blob/5b177803cb168dcaed0a8a6e2b2fbd835b82e0f9/src/websocket.js#L43
      isError: () => false, // Amberdata never receives error types?
      filter: (message: any) => !!message.params,
      toResponse: (message: any) => {
        const result = Requester.validateResultNumber(message, ['params', 'result', 'last'])
        return Requester.success('1', { data: { result } })
      },
    }
  }
}
