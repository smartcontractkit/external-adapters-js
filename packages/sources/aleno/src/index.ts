import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { price } from './endpoint'

export const adapter = new Adapter({
  defaultEndpoint: price.name,
  name: 'ALENO',
  config,
  endpoints: [price],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
