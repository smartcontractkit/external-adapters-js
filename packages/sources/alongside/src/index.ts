import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { Adapter } from '@chainlink/external-adapter-framework/adapter'
import { collateral } from './endpoint'
import { config } from './config'

export const adapter = new Adapter({
  defaultEndpoint: collateral.name,
  name: 'ALONGSIDE',
  config,
  endpoints: [collateral],
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
