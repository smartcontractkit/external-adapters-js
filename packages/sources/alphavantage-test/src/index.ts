import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { PriceAdapter } from '@chainlink/external-adapter-framework/adapter'
import { config } from './config'
import { forex } from './endpoint'

export const adapter = new PriceAdapter({
  defaultEndpoint: forex.name,
  name: 'ALPHAVANTAGE',
  config,
  endpoints: [forex],
  rateLimiting: {
    tiers: {
      free: {
        rateLimit1m: 5,
        rateLimit1h: 20.83,
      },
      '49.99': {
        rateLimit1m: 75,
      },
      '99.99': {
        rateLimit1m: 150,
      },
      '149.99': {
        rateLimit1m: 300,
      },
      '199.99': {
        rateLimit1m: 600,
      },
      '249.99': {
        rateLimit1m: 1200,
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
