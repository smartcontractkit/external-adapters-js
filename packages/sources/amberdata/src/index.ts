import { expose } from '@chainlink/ea-bootstrap'
import { makeExecute } from './adapter'
import { makeConfig, WSHandlerFactory, NAME } from './config'

export = { NAME, makeExecute, makeConfig, ...expose(makeExecute(), WSHandlerFactory()) }
