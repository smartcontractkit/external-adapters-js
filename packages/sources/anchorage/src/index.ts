import { expose, ServerInstance } from '@chainlink/external-adapter-framework'
import { config } from './config'
import { wallet } from './endpoint'
import { PoRAdapter } from '@chainlink/external-adapter-framework/adapter/por'

export const adapter = new PoRAdapter({
  defaultEndpoint: wallet.name,
  name: 'ANCHORAGE',
  config,
  endpoints: [wallet],
  rateLimiting: {
    tiers: {
      default: {
        rateLimit1m: 30,
        note: 'Docs: 10 requests per second per Organization, bursts of up to 100 requests per second',
      },
    },
  },
})

export const server = (): Promise<ServerInstance | undefined> => expose(adapter)
export const walletParameters = wallet.inputParameters
