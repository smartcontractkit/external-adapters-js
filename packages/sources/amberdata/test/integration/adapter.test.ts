import { AdapterRequest } from '@chainlink/ea-bootstrap'
import { server as startServer } from '../../src/index'
import {
  mockCryptoEndpoint,
  mockBalanceEndpoint,
  mockMarketCapEndpoint,
  mockVolumeEndpoint,
} from './fixtures'
import { setupExternalAdapterTest } from '@chainlink/ea-test-helpers'
import type { SuiteContext } from '@chainlink/ea-test-helpers'
import { SuperTest, Test } from 'supertest'

describe('amberdata', () => {
  const context: SuiteContext = {
    req: null,
    server: startServer,
  }

  const envVariables = {
    API_KEY: process.env.API_KEY || 'mock-api-key',
    CACHE_ENABLED: 'false',
  }

  setupExternalAdapterTest(envVariables, context)

  describe('when making a request to crypto endpoint', () => {
    const cryptoRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'crypto',
        base: 'ETH',
        quote: 'BTC',
      },
    }

    it('should reply with success', async () => {
      mockCryptoEndpoint()
      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(cryptoRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('when making a request to crypto endpoint with an override from config/overrides.json', () => {
    const cryptoRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'crypto',
        base: 'LUNA',
        quote: 'USD',
      },
    }

    it('should reply with success', async () => {
      mockCryptoEndpoint()
      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(cryptoRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('when making a request to marketcap endpoint', () => {
    const marketCapRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'marketcap',
        base: 'ETH',
      },
    }

    it('should reply with success', async () => {
      mockMarketCapEndpoint()
      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(marketCapRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('when making a request to volume endpoint', () => {
    const volumeRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'volume',
        base: 'LINK',
        quote: 'USD',
      },
    }

    it('should reply with success', async () => {
      mockVolumeEndpoint()
      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(volumeRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('when making a request to balance endpoint', () => {
    const balanceRequest: AdapterRequest = {
      id: '1',
      data: {
        endpoint: 'balance',
        addresses: [
          { address: '3EyjZ6CtEZEKyc719NZMyWaJpJG5jsVJL1' },
          { address: '38bzm6nhQMFJe71jJw1U7CbgNrVNpkonZF' },
        ],
        dataPath: 'addresses',
      },
    }

    it('should reply with success', async () => {
      mockBalanceEndpoint()
      const response = await (context.req as SuperTest<Test>)
        .post('/')
        .send(balanceRequest)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
