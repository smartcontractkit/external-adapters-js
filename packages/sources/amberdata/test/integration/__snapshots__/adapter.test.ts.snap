// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amberdata when making a request to balance endpoint should reply with success 1`] = `
{
  "data": {
    "result": [
      {
        "address": "3EyjZ6CtEZEKyc719NZMyWaJpJG5jsVJL1",
        "balance": "0",
        "chain": "mainnet",
        "coin": "btc",
      },
      {
        "address": "38bzm6nhQMFJe71jJw1U7CbgNrVNpkonZF",
        "balance": "2188",
        "chain": "mainnet",
        "coin": "btc",
      },
    ],
  },
  "jobRunID": "1",
  "result": [
    {
      "address": "3EyjZ6CtEZEKyc719NZMyWaJpJG5jsVJL1",
      "balance": "0",
      "chain": "mainnet",
      "coin": "btc",
    },
    {
      "address": "38bzm6nhQMFJe71jJw1U7CbgNrVNpkonZF",
      "balance": "2188",
      "chain": "mainnet",
      "coin": "btc",
    },
  ],
  "statusCode": 200,
}
`;

exports[`amberdata when making a request to crypto endpoint should reply with success 1`] = `
{
  "data": {
    "result": 0.06567038,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 0.06567038,
  "statusCode": 200,
}
`;

exports[`amberdata when making a request to crypto endpoint with an override from config/overrides.json should reply with success 1`] = `
{
  "data": {
    "result": 0.0000586300395535625,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 0.0000586300395535625,
  "statusCode": 200,
}
`;

exports[`amberdata when making a request to marketcap endpoint should reply with success 1`] = `
{
  "data": {
    "result": 391118025366,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 391118025366,
  "statusCode": 200,
}
`;

exports[`amberdata when making a request to volume endpoint should reply with success 1`] = `
{
  "data": {
    "result": 13804511.82541564,
  },
  "jobRunID": "1",
  "providerStatusCode": 200,
  "result": 13804511.82541564,
  "statusCode": 200,
}
`;
